src/net/socket_state.rs

```diff
@@ -120,20 +117,27 @@ impl SocketState {
 
                 // get specific slice of data for fragment
                 let fragment_data =
-                    &payload[start_fragment_pos as usize..end_fragment_pos as usize]; /* upcast is safe */
+                    &packet.payload()[start_fragment_pos as usize..end_fragment_pos as usize]; /* upcast is safe */
 
                 packet_data.add_fragment(&fragment, fragment_data.to_vec());
             }
         }
 
-        let mut lock = connection
-            .write()
-            .map_err(|error| NetworkError::poisoned_connection_error(error.description()))?;
+        let packet_addr = packet.addr();
```

Comment:
This line is only used once at line 140 so no extra allocation needed. Can get removed If you change line 140 to `packet.addr()`

Reply:
It cannot be removed since packet is effectively consumed at line 137. You would have a use after move error.